modules/enterprise/server/jar/src/main/java/org/rhq/enterprise/server/storage/StorageClientManager.java

```diff
@@ -631,4 +647,54 @@ public int getQueueAvailableCapacity() {
         return metricsServer.getQueueAvailableCapacity();
     }
 
+    /**
+     * A thread that checks for liveness of the given session.
+     */
+    private static class SessionAliveChecker extends Thread {
+
+        private static long SLEEP_TIME = 5000L;
+        private static long EXTENDED_SLEEP = SLEEP_TIME * 5;
+        private static int ALLOWED_FAILS = 2;
+
+        private final StorageClientManager storageClientManager;
+        private boolean alive = true;
+        private int fails = 0;
+
+        public SessionAliveChecker(StorageClientManager manager) {
+            this.storageClientManager = manager;
+        }
+
+        @Override
+        public void run() {
+            while(alive) {
+                try {
+                    Thread.sleep(SLEEP_TIME);
+                    try {
+                        storageClientManager.getMetricsDAO().checkLiveness(RHQ_KEYSPACE);
+                        if(fails > 0) {
+                            // Query succeeded, set fails to 0
+                            fails = 0;
+                        }
+                    } catch(QueryTimeoutException e) {
+                        LOG.error("Storage node connection check timed out");
+                    } catch(NoHostAvailableException e) {
+                        fails++;
+                        if(fails >= ALLOWED_FAILS) {
+                            LOG.error("Failed to contact the storage node for live check, recreating connection session");
+                            // We have lost the connection to the storage node, refresh and try again..
+                            storageClientManager.refreshSession();
+                            Thread.sleep(EXTENDED_SLEEP); // Sleep for a longer time to allow storage node to restart
+                        }
+                    }
+                } catch (InterruptedException e) {
+                    // Alive should be false in the next iteration if shutdown was called
+                }
```

Comment:
I would be inclined to add another catch block here for Exception or even possibly Throwable. If a RuntimeException is thrown for example, the SessionAliveChecker thread will die silently.